### ISSUE TITLE
Removing a namespace from OSM should disconnect proxies from that namespace

### ISSUE THREAD

**Issue by draychev** [openservicemesh/osm#4166]
When a namespace is removed all Envoys from that namespace should be disconnected.

This should be the case when we also apply an `ignore` label too.

**Comment by snehachhabria** [comment]
@mudit-01 sure, assigning the task to you

**Comment by johnsonshi** [comment]
Repro steps confirms that this is an issue.

![image](https://user-images.githubusercontent.com/13926417/151433482-88aca6af-e97d-412f-ba62-7cb9168c4d3d.png)

* As seen in the screenshot above, I ran the osm demo app at `./demo/run-osm-demo.sh`, which has 4 namespaces and 6 pods (with 6 Envoy proxies) under osm's supervision.
* Running `osm namespace remove bookbuyer` *correctly* causes the monitored namespace count to drop as expected, since osm should no longer monitor the bookbuyer namespace.
* However, the bookbuyer namespace contains 1 pod (1 Envoy proxy). As seen above, the count for connected Envoy proxies did NOT drop. *This is the bug.*

**Comment by johnsonshi** [comment]
Additional details:

When `osm namespace remove bookbuyer` is run, the single `bookbuyer` pod is actually NOT REMOVED from the mesh. This is incorrect behavior. This fact is confirmed when visiting the `osm-controller`'s `debug` server at `http://localhost:9092/debug/proxy`.

When `kubectl delete namespace bookstore` is run, the 2 `bookstore` pods are actually CORRECTLY REMOVED from the mesh. This is correct behavior. This fact is confirmed when visiting the `osm-controller`'s `debug` server.

As can be seen below, `kubectl delete namespace [ns]` causes proper deletion of a namespace (and its associated pods) from OSM. We can see that the `Connected Proxies` list and the `Connected Proxy Certificates list` (at `osm-controller`'s `debug` server at `http://localhost:9092/debug/proxy`) correctly do not include the 2 `bookstore` pods anymore.

As can be seen below, the `bookbuyer` pod is still connected because the pod's proxy cert list is still considered active and was not removed.

![image](https://user-images.githubusercontent.com/13926417/151452970-5d557cbc-52ff-405c-b0d7-bab3146cc778.png)

------

The bug seems to indicate that for `osm namespace remove`, the underlying pods' certs aren't deleted. Looking at `pkg/envoy/registry/announcement_handlers.go -> ReleaseCertificateHandler()`, we see that this announcement handler DOES remove Envoy proxy certs IF the `announcements.PodDeleted` happens.

We need to add a handler or modify this handler so that "pod removal events caused by `osm namespace remove`" properly removes the certs.

**Comment by johnsonshi** [comment]
From looking at `pkg/envoy/registry/announcement_handlers.go -> ReleaseCertificateHandler()`, osm only removes a pod from the mesh and "releases" the associated certificates when a `announcements.PodDeleted` event happens.

Looking at `pkg/k8s/client.go -> initPodMonitor()`, the `announcements.PodDeleted` event is only triggered and published on k8s pod deletion events.

This is unlike `pkg/k8s/client.go -> initNamespaceMonitor()`:

```
// Initializes Namespace monitoring
func (c *client) initNamespaceMonitor() {
	monitorNamespaceLabel := map[string]string{constants.OSMKubeResourceMonitorAnnotation: c.meshName}

	labelSelector := fields.SelectorFromSet(monitorNamespaceLabel).String()
	option := informers.WithTweakListOptions(func(opt *metav1.ListOptions) {
		opt.LabelSelector = labelSelector
	})

	informerFactory := informers.NewSharedInformerFactoryWithOptions(c.kubeClient, DefaultKubeEventResyncInterval, option)

	// Add informer
	c.informers[Namespaces] = informerFactory.Core().V1().Namespaces().Informer()

	// Add event handler to informer
	nsEventTypes := EventTypes{
		Add:    announcements.NamespaceAdded,
		Update: announcements.NamespaceUpdated,
		Delete: announcements.NamespaceDeleted,
	}
	c.informers[Namespaces].AddEventHandler(GetEventHandlerFuncs(nil, nsEventTypes, c.msgBroker))
}
```

As can be seen above, `announcements.NamespaceAdded` are triggered when the `constants.OSMKubeResourceMonitorAnnotation = meshName` label is added to namespaces.
Also, `announcements.NamespaceDeleted` are triggered for both K8S Namespace deletion events AND when the label is removed (but the namespace is not deleted).
This is the correct intended behavior, as a namespace is considered "deleted" from osm's point of view when the label is removed OR if k8s actually deletes the namespace.

As can be seen in `pkg/messaging/broker.go -> updateMetric()`, we decrement the namespace metric count when the `announcements.NamespaceDeleted` event happens.

---

In order to
1. properly delete the mesh certificate/actually kick out the pod's Envoy from the network and,
2. properly decrement the connected-pods metric,
we need to change the `initPodMonitor()` above so that:
1. a pod is considered "added" to the mesh (`announcements.PodDeleted` event) if it has the Envoy ID label set.
2. a pod is considered "deleted" from the mesh (`announcements.PodDeleted` event) if it has the Envoy ID label removed.

This will thus fix both (1) the certificate not getting released issue and (2) the connected-pod metric not being decremented issue. When the Envoy Unique ID Label is removed from the pod OR if k8s deletes the pod, for both events the `announcements.NamespaceDeleted` event happens.

When `announcements.NamespaceDeleted` event happens, `pkg/envoy/registry -> ReleaseCertificateHandler()` event handler gets triggered, which releases the certificate.

As can be seen in `pkg/envoy/ads/stream.go -> StreamAggregatedResources`, the method properly decreases the connected-pods metric (`metricsstore.DefaultMetricsStore.ProxyConnectCount.Dec()`) once the Envoy gRPC stream is closed (which will happen after the certificate is released by osm in `ReleaseCertificateHandler()` above.

**Comment by shashankram** [comment]
@johnsonshi This seems like a feature request to me. `osm namespace remove` was never meant to disconnect existing proxies from the mesh, but rather prevent new pods deployed in that namespace from being a part of the mesh.

So I want to start out by commenting that this is not a bug. If we want to support this, we should treat it as a feature request and evaluate the concrete requirements, as disconnecting proxies from the mesh requires internal control plane changes.

Before proceeding with the implementation, this feature requires a design doc that describes the requirements and high level design proposal in the CLI and controller.

**Comment by johnsonshi** [comment]
Thanks @shashankram. I design doc would make sense in this case. Thanks for letting me know that `osm namespace remove` was never meant to disconnect existing proxies from the mesh, but rather prevent new pods deployed in that namespace from being a part of the mesh. Yeah I'll outline all the proposed changes including documentation updates for existing behavior for clarity + my proposed feature changes.

**Comment by johnsonshi** [comment]
@shashankram I guess we also need to update our docs to actually outline the intended behavior of `osm namespace remove`, as it currently suggests existing in the namespace will be removed from the mesh.

```
$ bin/osm namespace remove --help

This command will remove a namespace from the mesh. All
services in this namespace will be removed from the mesh.
```

**Comment by shashankram** [comment]
> @shashankram I guess we also need to update our docs to actually outline the intended behavior of `osm namespace remove`, as it currently suggests existing in the namespace will be removed from the mesh.
> 
> ```
> $ bin/osm namespace remove --help
> 
> This command will remove a namespace from the mesh. All
> services in this namespace will be removed from the mesh.
> ```

@johnsonshi yes that comment in the help text is misleading.
We could rephrase it to:
```
This command will remove a namespace from the mesh. 
Existing pods belonging to the namespace must be restarted to 
remove the sidecar proxy injected by OSM.
```

I also think just disconnecting a proxy on `namespace remove` is not going to yield anything useful since the sidecar will still be on the pod. If users are expected to restart the pod anyway, I don't see much value in implementing this feature.

**Comment by steeling** [comment]
@shashankram good to close this? I don't think we should be in the business of restarting pods, and if that's the only solution to remove the sidecar then this seems like a non-starter

**Comment by shashankram** [comment]
> @shashankram good to close this? I don't think we should be in the business of restarting pods, and if that's the only solution to remove the sidecar then this seems like a non-starter

I am in favor of closing this as I don't believe we should be restarting user app pods to remove the sidecar when a namespace label is removed.

This can easily be achieved as follows:
```
osm namespace remove <ns>
kubectl rollout restart deploy <name> -n foo <ns>
```

**Comment by shashankram** [comment]
> > @shashankram good to close this? I don't think we should be in the business of restarting pods, and if that's the only solution to remove the sidecar then this seems like a non-starter
> 
> I am in favor of closing this as I don't believe we should be restarting user app pods to remove the sidecar when a namespace label is removed.
> 
> This can easily be achieved as follows:
> 
> ```
> osm namespace remove <ns>
> kubectl rollout restart deploy <name> -n foo <ns>
> ```

Also wanted to note that we could probably implement auto rejection of connections from proxies not a part of the mesh based on k8s label changes on namespaces/pods, however it would be easier simply to request the user to run the above 2 commands, as an unmanaged sidecar is the last thing a user would want.
